LinkControl: Add support for images in attachment search results

MaggieCabrera opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Closes #57776

Adds thumbnails for attachments that are images on link preview and the link control search list

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

An image gives you more info than the current icon for all attachments

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

By checking if the URL from the attachment is an image or not

## Screenshots or screencast <!-- if applicable -->

Attachments under the search result:

<img width="556" alt="Screenshot 2024-01-31 at 17 32 19" src="https://github.com/WordPress/gutenberg/assets/3593343/259007b2-047f-40ef-97b1-0231b8735a40">

Link preview of attachment that is NOT an image
<img width="466" alt="Screenshot 2024-01-31 at 17 31 52" src="https://github.com/WordPress/gutenberg/assets/3593343/f2bf8ec9-5176-491c-a1f3-cf1f68cfefcf">

Link preview of attachment that is an image
<img width="578" alt="Screenshot 2024-01-31 at 17 31 47" src="https://github.com/WordPress/gutenberg/assets/3593343/5a250b48-d4a8-48c6-9fe4-998561b15e31">

MaggieCabrera committed:
add image to link preview

MaggieCabrera committed:
added thumbnail to the search item

MaggieCabrera commented on packages/block-editor/src/components/link-control/link-preview.js:
```diff
@@ -56,8 +61,22 @@ export default function LinkPreview( {
+		icon = (
+			<img
+				className="block-editor-link-control__search-item-media-thumbnail"
+				src={ attachmentImg }
```
this loads the original sized image, it feels a little overkill

MaggieCabrera replied:
Both this and a thumbnail for the image are on coreStore, but the component shouldn't have to know about core-data. Should we pass that extra info to all instances of the link preview, or should we leave this as is?

getdave replied:
How big is the image? If it's huge then I don't think this is acceptable.

I did have a plan to use the `value.id` to fetch the full data about the entity from the Core Store. In fact I did have a PR about precisely that but I cannot locate it.

This selector would need to be passed in via blockEditorSettings because - as you noted - we cannot consume Core Store directly here.

If we do this I would consider whether it's better to do that bit as a separate PR.

MaggieCabrera replied:
https://github.com/WordPress/gutenberg/pull/50998 this one perhaps?

MaggieCabrera committed:
improved size and style of images

getdave commented on packages/block-editor/src/components/link-control/is-url-image.js:
```diff
@@ -0,0 +1,9 @@
+ * Checks if the url is an image.
+ * @param {string} url the url we are checking.
+ * @return {boolean} true if the url is an image url.
+ */
+export default function isImageUrl( url ) {
+	const pattern = /(http(s?):)([/|.|\w|\s|-])*\.(?:jpg|gif|png)/g;
+	return pattern.test( url );
+}
```
Some simple tests for this would be great.

MaggieCabrera replied:
I just saw that there is another PR (https://github.com/WordPress/gutenberg/pull/57824) open for the same thing and I prefer how they implement this check using mimetypes instead of a regex. I was going to grab that implementation myself

MaggieCabrera replied:
I spoke too soon! seems like it can be done for the suggestions because they come from `fetchLinkSuggestions` but the link previews don't have that, so we'd still have to figure out if the URL is an image via proding at the actual URL. I'll go about doing some tests for this function then.

getdave commented on packages/block-editor/src/components/link-control/link-preview.js:
```diff
@@ -56,8 +61,22 @@ export default function LinkPreview( {
 	if ( richData?.icon ) {
 		icon = <img src={ richData?.icon } alt="" />;
+	} else if ( attachmentImg ) {
+		//TODO we don't have an alt text for this image
```
An empty `alt` tag is acceptable when there is no way to provide an accurate description

getdave reviewed (commented):
Thanks for working on this.

Some things I noticed.

MaggieCabrera commented:
@getdave the image is as big as the original image is. In my local env I was using huge images for some other reason and I was seeing a 2k pixel width image in the tiny thumbnail. It could definitely be a followup

MaggieCabrera committed:
added unit test for isImageUrl

MaggieCabrera commented:
> > @getdave the image is as big as the original image is. In my local env I was using huge images for some other reason and I was seeing a 2k pixel width image in the tiny thumbnail. It could definitely be a followup
> 
> If I've understood you correctly, I don't think we should introduce the _possibility_ of loading giant images in a preview for a link.

I might have not been clear, this is what I can see in my test site

<img width="681" alt="Screenshot 2024-02-01 at 16 21 45" src="https://github.com/WordPress/gutenberg/assets/3593343/f4d742bf-cf8a-470e-aee8-66a87e7ae51a">

draganescu commented:
This PR is cool in terms of improving the UI of adding links. i think for sighted users the small area of preview can speed up choosing the image picking process of choosing from the suggestions or search results.

However, we need to show a really tiny thumbnail there not the full image. Can't we get the thumbnail via another [REST call](https://developer.wordpress.org/rest-api/reference/media/) and replace async the previews with what we receive from the server?

Also for unsighted users the Alt can say "Thumbnail preview of [filename]".

getdave commented:
As @draganescu indicated, we can alter fetchLinkSuggestions to query the Media endpoint rather than search. We would need to be sure to continue to offer the same properties on the returned object but we might get access to more data about the media which we could use to _augment_ the response.

MaggieCabrera committed:
Includes a new REST API endpoint for searching media items

Co-authored-by: petitphp <petitphp@git.wordpress.org>

MaggieCabrera committed:
used thumbnail for search suggestion

MaggieCabrera commented on lib/compat/wordpress-6.6/class-wp-rest-media-search-handler-gutenberg-6-6.php:
```diff
@@ -0,0 +1,110 @@
+<?php
+/**
+ * REST API: WP_REST_Media_Search_Handler class
+ *
+ * @package WordPress
+ * @subpackage REST_API
```
I replicated the intention of [this PR](https://github.com/WordPress/gutenberg/pull/58638) and gave credit to the author in case this one gets merged first

MaggieCabrera committed:
removed logs

MaggieCabrera committed:
make sure imageURL is not empty

MaggieCabrera committed:
added alt text

MaggieCabrera commented:
I added the new endpoint and added a thumbnail and alt text to it. Should we remove the link preview implementation from this PR and do that in a follow-up to keep this one simpler?

MaggieCabrera committed:
renamed class to follow convention

MaggieCabrera committed:
linting errors
